### DISCUSSION TITLE
How to create a WinUI3 application that has no title bar and no minimize, maximize, and close buttons?

### DISCUSSION THREAD

**Discussion by stefanvd** [microsoft/microsoft-ui-xaml#8800] [Q&A] [answered]
Hi there,

I am creating a WinUI3 desktop application that has its own custom close action button in the center of the Mainwindow. Additionally, I want to remove the entire title bar along with the minimize [-], maximize [ ], and close [X] controls. I use the _ExtendViewIntoTitleBar_ but that does not remove the window control buttons. And that keeps it window border-radius and shadow. Furthermore, I do not want the application to appear in the Windows taskbar as an extra application.
Could you please provide a step-by-step description of how to achieve this?

**Comment by castorix** [comment]
You can use a Layered window (like [WinUI3_SwapChainPanel_Layered](https://github.com/castorix/WinUI3_SwapChainPanel_Layered))
To remove the app from the Taskbar, there is [AppWindow.IsShownInSwitchers](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.windowing.appwindow.isshowninswitchers?view=windows-app-sdk-1.2) (which calls [ITaskbarList::DeleteTab](https://learn.microsoft.com/en-us/windows/win32/api/shobjidl_core/nf-shobjidl_core-itaskbarlist-deletetab))

**Reply by stefanvd** [reply]
Is there no simpler version? It doesn't even work here on my Windows 11.

**Reply by DarranRowe** [reply]
For the title bar. The question is whether you want the window to be resizable or not.
If you don't want the window to be resizable, then it is relatively simple.

![Screenshot 2023-08-30 174152](https://github.com/microsoft/microsoft-ui-xaml/assets/52577874/06829ae5-4d68-4e55-93e3-588133ddffdc)

```
    MainWindow::MainWindow()
    {
        InitializeComponent();

        auto wn = try_as<IWindowNative>();

        HWND window_handle{};
        wn->get_WindowHandle(&window_handle);

        auto presenter = AppWindow().Presenter().as<Microsoft::UI::Windowing::OverlappedPresenter>();
        presenter.IsMaximizable(false);
        presenter.IsMinimizable(false);
        presenter.IsResizable(false);
        presenter.SetBorderAndTitleBar(true, false);

        DWM_WINDOW_CORNER_PREFERENCE wcp = DWMWCP_ROUND;
        DwmSetWindowAttribute(window_handle, DWMWA_WINDOW_CORNER_PREFERENCE, &wcp, sizeof(DWM_WINDOW_CORNER_PREFERENCE));
    }
```

You use the presenter to disable the title bar, and then use DwmSetWindowAttribute to enable the rounded corners. Unfortunately, this is required.

Unfortunately, if you want the window to be resizable then there is a long running Windows problem.
```
    MainWindow::MainWindow()
    {
        InitializeComponent();

        auto presenter = AppWindow().Presenter().as<Microsoft::UI::Windowing::OverlappedPresenter>();
        presenter.IsMaximizable(false);
        presenter.IsMinimizable(false);
        presenter.SetBorderAndTitleBar(true, false);
    }
```
![Screenshot 2023-08-30 174604](https://github.com/microsoft/microsoft-ui-xaml/assets/52577874/b3c0c459-f7f5-4f3e-b8b1-f5f97cea6463)

There is a small region at the top of the window which looks like Windows is drawing part of the title bar. If this is unacceptable, your only two options are to explore handling the non client messages to see if you can remove this, or use DwmExtendFrameIntoClientArea to take complete control of the border. There is also the option of setting up a campaign to get lots of people to report/upvote this problem on the feedback hub to try to get Microsoft to fix this in Windows.

Oh, and as an FYI, the "App title" text is from a WinUI text block.

**Reply by stefanvd** [reply]
Thank you for your answer. I do not need to resize the window; it is always a fixed size. However, is that C# code? I am a beginner developer and use C# in my blank WinUI3 application.

**Reply by DarranRowe** [reply, marked as answer]
It is C++, but it is possible to use this same functionality from C#. WinUI is based upon the same set of underlying components, so most of that should work as is. The biggest issue is with the call to DwmSetWindowAttribute, since that would have to be a platform invoke call.
This would mean that it is pretty trivial to convert between the two when working with the WinUI code, and WinRT component code in general.

```
namespace CsMeh
{
    public class NativeApi
    {
        //Platform invoke definition for DwmSetWindowAttribute
        [DllImport("dwmapi.dll")]
        public static extern int DwmSetWindowAttribute(IntPtr hwnd, int attribute, ref int value, int size);
    }

    public sealed partial class MainWindow : Window
    {
        public MainWindow()
        {
            this.InitializeComponent();

            var presenter = AppWindow.Presenter as OverlappedPresenter;
            presenter.IsMaximizable = false;
            presenter.IsMinimizable = false;
            presenter.IsResizable = false;
            presenter.SetBorderAndTitleBar(true, false);

            //This Win32Interop is in the Microsoft.UI namespace.
            var windowHandle = Win32Interop.GetWindowFromWindowId(AppWindow.Id);

            //DWM_WINDOW_CORNER_PREFERENCE documentation gives DWMWCP_ROUND as 2
            int cornerPreference = 2;
            //DWMWA_WINDOW_CORNER_PREFERENCE is documented to have a value of 33
            NativeApi.DwmSetWindowAttribute(windowHandle, 33, ref cornerPreference, Marshal.SizeOf<int>());
        }
    }
}
```
This is similar code translated to C#. I have to admit that using IWindowNative in the C++ code is habitual now.
The [Window](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.window?view=windows-app-sdk-1.3) documentation states that there is an [AppWindow](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.windowing.appwindow?view=windows-app-sdk-1.3) property. This in turn states that it has a Presenter property. The presenter for a top level window like this should always be an [OverlappedPresenter](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.windowing.overlappedpresenter?view=windows-app-sdk-1.3), which is why there is that immediate conversion. AppWindow.Presenter returns the base, AppWindowPresenter, by default. If you want to be safe, [AppWindowPresenter](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.windowing.appwindowpresenter?view=windows-app-sdk-1.3) has the Kind property. This allows you to verify that it is an overlapped presenter.

The next portion of this is the call to [DwmSetWindowAttribute](https://learn.microsoft.com/en-us/windows/win32/api/dwmapi/nf-dwmapi-dwmsetwindowattribute). This is a Windows API function and I honestly don't know if there is an alternative. I am a C++ programmer, so I am not very good with C#. The first parameter of DwmSetWindowAttribute is a handle to the window that you want to work on. To get this handle, you can use one of two methods. There is the method used in the code above, where I query AppWindow for its ID, and then I use GetWindowFromWindowId to get the handle. The other option is:
```
var windowHandle = WinRT.Interop.WindowNative.GetWindowHandle(this);
```
This is the equivalent of using the IWindowNative interface in my C++ example.
It is up to you which you want to use, but please remember that both of these can only be used from a class that derives from Window. If you are working with a user control or a page, then this needs to be done for the containing window. Anyway, this handle is how the Windows API identifies a window.
Once you have this handle then it is easy to call DwmSetWindowAttribute. The values were obtained from the documentation. [DWMWINDOWATTRIBUTE](https://learn.microsoft.com/en-us/windows/win32/api/dwmapi/ne-dwmapi-dwmwindowattribute) documents DWMWA_WINDOW_CORNER_PREFERENCE to be 33. It also states that the value should be from the [DWM_WINDOW_CORNER_PREFERENCE](https://learn.microsoft.com/en-us/windows/win32/api/dwmapi/ne-dwmapi-dwm_window_corner_preference) enumeration. This enumeration states that DWMWCP_ROUND has a value of 2. So this is where these values came from.

I obtained the platform invoke signature from [pinvoke.net](https://www.pinvoke.net/default.aspx/Enums/DwmSetWindowAttribute.html). If I was being thorough, then I would have properly defined all of the enumerations. For example, [DWMWINDOWATTRIBUTE](http://www.pinvoke.net/default.aspx/Enums/DWMWINDOWATTRIBUTE.html) and DWM_WINDOW_CORNER_PREFERENCE, but this was fine for a sample.

**Reply by stefanvd** [reply]
Thank you so much! I finally get the result I want to see in my WinUI3 app window.
